fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectory.java

```diff
@@ -596,25 +597,36 @@ public void deleteFile(@Nonnull String name) throws IOException {
             return;
         }
         try {
-            boolean deleted = Objects.requireNonNull(context.asyncToSync(LuceneEvents.Waits.WAIT_LUCENE_DELETE_FILE, getFileReferenceCacheAsync().thenApply(cache -> {
-                FDBLuceneFileReference value = cache.get(name);
-                if (value == null) {
-                    return false;
-                }
-                context.ensureActive().clear(metaSubspace.pack(name));
-                context.ensureActive().clear(dataSubspace.subspace(Tuple.from(value.getId())).range());
-                cache.remove(name);
-                return true;
-            })));
+            boolean deleted = Objects.requireNonNull(context.asyncToSync(LuceneEvents.Waits.WAIT_LUCENE_DELETE_FILE, getFileReferenceCacheAsync()
+                    .thenApply(cache -> deleteFileInternal(cache, name))));
 
             if (!deleted) {
                 throw new NoSuchFileException(name);
             }
+
+            if (isCompoundFile(name)) {
+                Map<String, FDBLuceneFileReference> cache = fileReferenceCache.get();
+                String primaryKeyName = name.substring(0, name.length() - DATA_EXTENSION.length()) + "pky";
+                deleteFileInternal(cache, primaryKeyName);
+                // TODO: If the segment is being deleted because it no longer has any live docs, it won't be merged
+                //  so we might need to delete the primary key entries another way for case where tryDeleteDocument
+                //  wasn't used.
+            }
         } finally {
             context.increment(LuceneEvents.Counts.LUCENE_DELETE_FILE);
         }
     }
 
+    private boolean deleteFileInternal(@Nonnull Map<String, FDBLuceneFileReference> cache, @Nonnull String name) {
+        FDBLuceneFileReference value = cache.remove(name);
+        if (value == null) {
+            return false;
+        }
+        context.ensureActive().clear(metaSubspace.pack(name));
+        context.ensureActive().clear(dataSubspace.subspace(Tuple.from(value.getId())).range());
```

Comment:
Are these valid? Do we store anything for `*.pky` in the actual `dataSubspace`?

Reply:
We do not store anything there now; should we just skip this clear?